### ISSUE TITLE
feat(inference): postpay credit mode for router aggregator — defer top-up, reuse existing settlement path

### ISSUE THREAD

**Issue by Ravenyjh** [0glabs/0g-serving-broker#474]
## Background

0G runs a router that fronts inference requests on behalf of end users across many providers. Under the current prepay model, the router must pre-fund every provider's sub-account, locking up working capital proportional to `forecast × N_providers`. This is prohibitive at scale.

What the router needs: **send requests on credit, with a hard per-provider cap, and have payment flow through the existing on-chain settlement path** — without any contract changes.

> Audit-mode settlement (pre-ack provider verification by 0G) is **out of scope here** — see #475. That use case requires bypassing the on-chain `Acknowledged` signal, which is a different concern from capital efficiency. The two mechanisms compose but are independently scoped.

### Core idea

Keep the existing `depositFund` + `SettleFeesWithTEE` contract path intact. Move the **timing** of `depositFund` from "before usage" to "after usage, triggered by accrued debt":

```
Current (prepay):  depositFund → accrue usage → SettleFeesWithTEE
Proposed (postpay): accrue usage (on credit, capped) → depositFund → SettleFeesWithTEE
```

**Zero contract changes.** All work happens in broker.

### Flow

| Step | Actor | Action |
|---|---|---|
| T0 | Router | One-time `acknowledgeProviderSigner` per provider (no capital required). For router this is fine — 0G operates the router and ack-ing reflects legitimate operational endorsement. |
| T1 | Router → Broker | Sends request; broker matches postpay whitelist, **skips balance validation** in `ValidateRequestWithEstimatedFee`, keeps all other checks (session sig, `Acknowledged`, `TeeSignerAcknowledged`) |
| T2 | Broker | Records request in DB; tracks per-address accrued unsettled fee |
| T2.5 | Broker | If `accruedDebt + estimatedFee > maxUnsettledFee` cap → reject new requests (429) until top-up clears debt |
| T3 | Broker | When `accruedDebt > topUpThreshold` → notify router (webhook / pull endpoint / on-chain event) |
| T4 | Router | Calls existing `depositFund(provider, amount)` for the accrued amount |
| T5 | Broker | Existing `ProcessSettlement` runs `SettleFeesWithTEE` — funds flow to provider on-chain as today |

### Why this preserves decentralization

- Provider receives funds via the existing on-chain `SettleFeesWithTEE` path — same TEE signature verification, same nonce protection, same balance accounting.
- Cap is enforced **automatically** by the broker (not by manual review). Maximum provider risk exposure = `maxUnsettledFee`, deterministic and bounded.
- Cap and webhook URL are configured per-provider — each provider decides whether and how much credit to extend to a given postpay address. Not a global trust.
- No off-chain billing rail introduced; no provider must "trust 0G to pay via fiat".

### Capital efficiency

Steady-state working capital for router drops from `~forecast × N_providers` (prepay) to `~peak_in_flight_debt` (postpay). Order-of-magnitude estimate: 10–30× reduction depending on usage distribution and top-up cadence.

### Config (new — separate from existing `WhitelistConfig`)

```yaml
postpay:
  enabled: true
  accounts:
    - address: "0xRouterAddr"
      maxUnsettledFee: "5000000000000000000"   # 5 token; hard cap
      topUpThreshold: "1000000000000000000"    # 1 token; trigger notification
      notification:
        mode: "webhook"                         # webhook | pull | event
        webhookUrl: "https://router.0g.ai/postpay/notify"
```

### Code changes

| File | Change |
|---|---|
| `config/config.go` | Add `PostpayConfig` (separate from `WhitelistConfig`) |
| `internal/ctrl/ctrl.go` | Load postpay accounts; expose `IsPostpayUser(addr)`, `GetPostpayCap(addr)` |
| `internal/ctrl/request.go:266` (`ValidateRequestWithEstimatedFee`) | For postpay users: skip `validateBalanceAdequacy`; add cap check on `accruedDebt + estimatedFee` |
| `internal/ctrl/settlement_tee.go` | After computing per-user accrued fees, emit top-up notification when threshold crossed (idempotent — track `last_notified_at`) |
| `internal/handler/postpay.go` (new) | `GET /v1/postpay/balance-due?address=X` for pull-mode routers |
| `internal/proxy/proxy.go` | Tag postpay requests in metrics (`payment_mode=postpay`) |
| `monitor/` | New gauge: per-address accrued unsettled fee; alerting on cap-approach |

**No new DB tables required** — `Request` table already tracks unsettled fees per user; aggregation via existing `ListUsersWithUnsettledFees` query.

**No contract changes.**

### Notification mechanism — pick one

- **Webhook (push)**: simplest, suitable when 0G operates the router. Broker calls configured URL with `{address, accruedFee, providerAddr}`.
- **Pull endpoint**: most robust — broker has no outbound dependency. Router polls `/v1/postpay/balance-due` periodically.
- **On-chain event**: most decentralized; requires watcher infrastructure on router side.

Recommend supporting **webhook + pull** in v1; on-chain event optional later.

## Composition with audit-mode (#475)

These mechanisms compose cleanly but are independently scoped:

- **Router (this issue)**: keeps `Acknowledged`, gets capital efficiency via deferred top-up.
- **Audit (#475)**: bypasses `Acknowledged` via new contract method `settleFromAuditor`, avoids the on-chain trust signal.
- **Auditor + postpay (both)**: an auditor address can be registered in BOTH `PostpayConfig` (broker) AND `auditorRegistry` (contract) to get pre-ack access AND no pre-fund. Settlement-path selection is per-user-role at settlement time.

This issue is the prerequisite-free piece — ship it first; audit-mode follows on its own contract upgrade timeline.

## Out of scope

- Router-side top-up bot implementation (separate microservice, but design should be documented alongside).
- Migration from existing free `WhitelistConfig` — keep both mechanisms, document when to use which.
- Multi-tenant / per-end-user accounting on the router side (router's internal concern, not broker's).
- Audit-mode settlement (see #475).

## Risks & mitigations

| Risk | Mitigation |
|---|---|
| Router fails to top up after notification | Cap auto-stops new requests; provider's max loss = `maxUnsettledFee` |
| Webhook delivery fails | Retry with backoff; fall back to pull mode if configured; log to alerting |
| Broker accrues runaway debt due to bug in cap check | Add invariant check in settlement loop; alert if debt > 1.5× cap |
| Confusion with existing free whitelist | Separate config struct, separate metrics label, doc in CLAUDE.md / design doc |
| TEE verification weakened | **Not weakened** — `Acknowledged` and `TeeSignerAcknowledged` checks are preserved; only balance check is skipped |

## Acceptance criteria

- [ ] Postpay-whitelisted address can send inference requests without pre-funded sub-account balance
- [ ] Per-address `maxUnsettledFee` cap enforced — requests rejected with 429 once exceeded
- [ ] Top-up notification fires when `accruedDebt > topUpThreshold` (webhook + pull endpoint both supported)
- [ ] After router calls existing `depositFund`, normal `SettleFeesWithTEE` settlement runs without modification
- [ ] Metrics distinguish `payment_mode=postpay` from `prepay` and `whitelist` (free)
- [ ] Existing prepay flow and free whitelist behavior unchanged
- [ ] Integration test: postpay user → request → cap enforcement → manual top-up → settlement → fee landed at provider on-chain

## Discussion points

- What's the right default `topUpThreshold` ratio relative to `maxUnsettledFee`? (Suggest 20% — gives router 5× the threshold-to-cap window for top-up to land.)
- Should broker auto-trigger a `forceSettlement` after detecting top-up, or wait for the regular settlement loop?
- Notification mechanism preference — webhook vs pull — for the v1 0G router rollout.

Related: #475 (audit-mode settlement) — composable, not blocking.
